pkg/controller/builder.go

```diff
@@ -219,34 +222,56 @@ func setContainerEnv(con *corev1.Container, envs []corev1.EnvVar) error {
 
 func UpdateDeploymentProxyHook(pc clients.ProxyClientInterface) deploymentcontroller.DeploymentHookFunc {
 	return func(_ *operatorv1.OperatorSpec, deployment *appsv1.Deployment) error {
-		klog.FromContext(context.Background()).WithName("builder").V(0).Info("Updating environment", "deployment", deployment.Name)
+		klog.FromContext(context.Background()).WithName("builder").V(0).Info("ProxyHook updating environment", "deployment", deployment.Name)
 		proxyConfig, err := pc.Get("cluster")
 		if err != nil {
 			return fmt.Errorf("error getting proxies.config.openshift.io/cluster: %w", err)
 		}
 
-		var errs []error
 		vars := []corev1.EnvVar{
 			{Name: HTTPSProxy, Value: proxyConfig.Status.HTTPSProxy},
 			{Name: HTTPProxy, Value: proxyConfig.Status.HTTPProxy},
 			{Name: NoProxy, Value: proxyConfig.Status.NoProxy},
 		}
+		return setDeploymentContainersEnvVars(vars, deployment)
+	}
+}
 
-		for i := range deployment.Spec.Template.Spec.InitContainers {
-			err = setContainerEnv(&deployment.Spec.Template.Spec.InitContainers[i], vars)
-			if err != nil {
-				errs = append(errs, err)
-			}
+func UpdateDeploymentFeatureGatesHook(fa featuregates.FeatureGateAccess) deploymentcontroller.DeploymentHookFunc {
+	return func(_ *operatorv1.OperatorSpec, deployment *appsv1.Deployment) error {
+		klog.FromContext(context.Background()).WithName("builder").V(0).Info("FeatureGatesHook updating environment", "deployment", deployment.Name)
+
+		featuregates, err := fa.CurrentFeatureGates()
+		if err != nil {
+			return fmt.Errorf("error getting featuregates.config.openshift.io/cluster: %w", err)
 		}
-		for i := range deployment.Spec.Template.Spec.Containers {
-			err = setContainerEnv(&deployment.Spec.Template.Spec.Containers[i], vars)
-			if err != nil {
-				errs = append(errs, err)
-			}
+
+		vars := []corev1.EnvVar{{Name: "FEATURE_GATES"}}
```

Comment:
Does the upstream feature gate code know about the `FEATURE_GATES` env? I have always used the `--feature-gates` flag upstream.

Reply:
that's a good question.
Just on a face value, if `--feature-gates` arg is supported, afaik most of the libraries should be able to handle env var form as well, but I will have to check that upstream to be sure.
If it won't or there won't be too much work required to change this to an arg, I'll do that since it seems it's usually passed this way.

Reply:
I switched to args and mapped to only one valid upstream feature gate which seems to be accepted (even though AFAIR it has been recently removed from upstream upstream ;)) and was able to successfully validate the behavior together with manager actually reading the feature gates that are set